include/xtensor-signal/find_peaks.hpp

```diff
@@ -0,0 +1,725 @@
+                xt::xarray<size_t> left_edges = xt::adapt(_left_edges, { _left_edges.size() });
+                xt::xarray<size_t> right_edges = xt::adapt(_right_edges, { _right_edges.size() });
+                return std::make_tuple(midpoints, left_edges, right_edges);
+            }
+        }
+
+        /**
+         * @brief implements the peak widths interface.
+         * @param x 1D data vector which matches the data vector used to generate peaks.
+         * @param peaks locations of maximums in x. Generally generated from find peaks.
+         * @param rel_height defines the height of the peak that is used to generate the width.
+         * @param wlen defines the window length used to calculate prominance.
+         * @returns tuple of arrays of widths, width heights, left bound and right bounds.
+         */
+        template<
+            typename E1,
+            typename E2,
+            typename E3 = double>
+            auto peak_widths(
+                E1&& x,
+                E2&& peaks,
+                E3&& rel_height,
+                std::optional<size_t> wlen)
+        {
+            //check that we have only one dimention
+            if (x.shape().size() != 1)
+            {
+                throw std::runtime_error(
+                    "Array must be 1D");
+            }
+
+            if (peaks.shape().size() != 1)
+            {
+                throw std::runtime_error(
+                    "Array must be 1D");
+            }
+
+            //declare an internal variable for prominance data
+            xt::xarray<double> _prominence_data;
+            //check if prominance data is being supplied
+
+            //Calculate prominence if not supplied and use wlen if supplied.
+            //check if wlen is an acceptable parameter
+            auto wlen_safe = detail::arg_wlen_as_expected(wlen);
+            auto [prominences, left_bases, right_bases] = detail::peak_prominences(x, peaks, wlen_safe);
+            return detail::peak_widths(x, peaks, rel_height, prominences, left_bases, right_bases);
+        }
+
+        /**
+         * @brief finds all peaks and applies filters to the peaks based on the parameters provided.
+         * @details portions of this function are unimplemented and will throw a compiler error if features requested are not available.
+         * @param 1D data vector.
+         * @param minimum height thresholds to be filtered. Only minimum heights are implemented.
+         * @param threshold unimplemented.
+         * @param distance unimplemented.
+         * @param prominence unimplemented.
+         * @param width defines the minimum width for eligable points. if a second parameter is applied a max width is also applied.
+         * @param wlen defined wlen for calculating prominence.
+         * @param rel_height used in defining the width of peaks based on the location. Defaults to .5  or 6dB.
+         * @param plateau_size unimplemented.
+         */
```

Comment:
The `param` arguments in this doc do not really make sense for a class. This should be splitted accross the different setters. A "summary" can still be in the description of the class, but not as `@param` (this would probably result in a doxygen / breathe error).